### ISSUE TITLE
Suboptimal execution order of parallel map calls for tf.data

### ISSUE THREAD

**Issue by eriikj** [tensorflow/tensorflow#39992]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 and Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2
- Python version: 3.6.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
When using a num_parallel_calls larger than the number of worker threads in the threadpool in a Dataset.map call, the order of execution is more or less random, causing a busty output behavior.

If the dataset map transform has a list of 20 elements to process, it typically processes them in a order that looks something like this:
5, 4, 7, 6, 1, 0, 2, 3, 18, 15, 10, 9, 13, 19, 14, 8, 17, 10, 12, 11

This is problematic since the output has to be contiguous, so no output will be available until a large portion of the calls in the threadpool have been processed.

I have attached a file with a self contained example which reproduces this behavior.

There are workarounds, such as allowing non-deterministic output, but in the long run, we want our trainings to be as deterministic as possible to aid debugging, so fixing this behavior would be very helpful for us.

**Describe the expected behavior**
I expect the map call to start processing the next unprocessed element in the dataset whenever it has a free worker thread, so the results can be made available as soon as possible.

**Standalone code to reproduce the issue**
See attached file.
[parallel_map_test.zip](https://github.com/tensorflow/tensorflow/files/4702027/parallel_map_test.zip)

**Other info / logs**
Example output from the code is also available in the attached file.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/8471d51d7c8204008b6d6e62726ae9bf/39992.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/1a32d5ca9f34b8dcb1b68916d3430cc6/39992-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by jsimsa** [comment]
@eriikj tf.data uses Eigen's threadpool [under the hoods](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/threadpool.cc#L96-L106), which does not guarantee FIFO scheduling. Switching to a different threadpool implementation and/or implementing FIFO ordering within tf.data would be non-trivial and it is not clear to me that it would be worthwhile.

As per tf.data [performance guide](https://www.tensorflow.org/guide/data_performance), users are recommended to use `tf.data.experimental.AUTOTUNE` for `num_parallel_calls` and leave the determination of the optimal value to the tf.data runtime. It is much more likely that the tf.data team will prioritize improving the autotuning logic over improving performance of misconfigured input pipelines.

**Comment by jsimsa** [comment]
Sorry, didn't mean to close this.

**Comment by eriikj** [comment]
I'm not sure why you consider this to be misconfigured, but fair enough. I will try using autotune and see what happens. Is it ok to keep the issue open until I have tested this?

As a side note: autotune would likely be able to do a better job if it didn't have to deal with this problem, so when improving autotune, it might be worthwhile to keep this in mind.

**Comment by zhuzilin** [comment]
@eriikj 
I think using autotune is the same as seting a lower num_parallel_calls, because the inner logic of autotune in ParallelMapDataset is:
```c++
      if (num_parallel_calls_->value == model::kAutotune) {
        num_parallel_calls_->value = ctx->runner_threadpool_size();
      }
```
Any reason why you need a num_parallel_calls that is much larger than worker threads in the threadpool?

**Comment by eriikj** [comment]
@zhuzilin 
If that's how autotune works, using it will cause another problem. Our first approach was to just use num_parallel_calls=8 (which is the same as the threadpool size), but we found that we did not have the same performance as before using the tf.data API (we are currently using a Keras Sequence as input, which handles the queuing as expected in these cases).

My first investigation found that setting num_parallel_calls to 8 implicitly creates a queue limited to the same size 8. Every started call holds a place in the queue until it can be sent as output. Since the output still has to be in order, this might be at a time later than when the call is completed. This causes the queue to sometimes be full, even though there are worker threads without work, causing a loss in performance. I thought that increasing the number of parallel calls, but keeping the number of worker threads low would solve this, and that's when I ran into the problem originally described in this issue.

I have attached a modified variant of the example I provided earlier with the same problems we saw with that configuration (with a little more variance in the sample fetch time to amplify the problem).

[parallel_map_test_num_parallel_calls_8.zip](https://github.com/tensorflow/tensorflow/files/4723607/parallel_map_test_num_parallel_calls_8.zip)

Here is a part of the output from the script which illustrates the problem that some worker threads have to wait (in this case more than a second) to start processing a new call:

2020-06-03 15:15:58.447852: I tensorflow/core/kernels/logging_ops.cc:166] started 31
2020-06-03 15:15:58.456823: I tensorflow/core/kernels/logging_ops.cc:166] stopped 26
2020-06-03 15:15:58.505422: I tensorflow/core/kernels/logging_ops.cc:166] stopped 28
2020-06-03 15:15:58.571174: I tensorflow/core/kernels/logging_ops.cc:166] stopped 30
2020-06-03 15:15:58.909244: I tensorflow/core/kernels/logging_ops.cc:166] stopped 25
2020-06-03 15:15:59.413419: I tensorflow/core/kernels/logging_ops.cc:166] stopped 24
2020-06-03 15:15:59.609620: I tensorflow/core/kernels/logging_ops.cc:166] stopped 23
2020-06-03 15:15:59.610705: I tensorflow/core/kernels/logging_ops.cc:166] output available: 23
2020-06-03 15:15:59.611575: I tensorflow/core/kernels/logging_ops.cc:166] started 32

**Comment by jsimsa** [comment]
My definition for misconfiguration is to be using a fixed value of `num_parallel_calls` that results in sub-optimal performance.

I do not follow your explanation for why using `num_parallel_calls` that matches the size of the threadpool did not work for you. I am also not sure how to interpret your comment about the use of Keras Sequence API; are you currently using Keras Sequence APIs and are in the process of switching to tf.data? or have you used tf.data and because of issue with `num_parallel_calls` switched to using Keras Sequence API?

As for parallel map internals. The `num_parallel_calls` argument controls two things: 1) the maximum degree of parallelism and the size of internal buffer used for storing the results. If the buffer is full, the no further computation is performed. If you find the depth of the buffer insufficient (because of variance in the speed of produced and consumer), you should add `prefetch` to your input pipeline (as opposed to increasing the size of `num_parallel_calls` beyond the size of the threadpool).

**Comment by eriikj** [comment]
We are currently using the Keras Sequence API and trying to migrate to the tf.data API.

We are limited by our data pipeline, prefetch does not do anything for us, since all elements are consumed almost immediately.

I have provided some code that should illustrate the problem with setting num_parallel_calls to the same as the number of threads in the pool, but I will try to explain the problem as well:

If we have 8 worker threads and num_parallel_calls set to 8, we will have a internal buffer of size 8 as well. If we have 8 calls (call 0 to 7) running in 8 threads and any call after call 0 finishes before call 0. the results of the finished calls will occupy a position in the buffer until all calls before it finishes. This causes the that thread to sit idle during this time.

Here is an example of when worker threads are idle, even though the dataset has more items to process:

initial state: 0, 1, 2, 3, 4, 5, 6, 7 running
call 5 finishes: 0, 1, 2, 3, 4, 6, 7 running, the results of call 5 still in the buffer (thread 5 is idle)
call 6 finishes: 0, 1, 2, 3, 4, 5, 6, 7 running, the results of call 5 and 6 still in the buffer (thread 5 and 6 are idle)
call 0, 1, 2, 3 and 4 finishes: 7, 8, 9, 10, 11, 12, 13, 14 running, results of call 0, 1, 2, 3, 4, 5 and 6 is sent as output

This would also apply to the case with autotune, if it works as stated in the comments above (that it simply sets num_parallel_calls to the number of worker threads). It's possible that better autotune logic might work around this problem, but I think it would be easier to address the cause of the problem. I realize that the change might not be trivial (I actually had a go at it before writing this issue, but came to the same conclusion), so I guess it comes down to a question of priority.

**Comment by jsimsa** [comment]
Thank you for your explanation. I acknowledge the problem. A possible solution for it would be to decouple the configuration of parallelism from the configuration of the buffering. I was hoping to achieve using existing building blocks through my suggestion to use `prefetch` but I realized it would not work because of head-of-the-line blocking. 

As you pointed out, it is possible to avoid this issue by using non-determinism. What you are requesting is to improve the performance of deterministic parallel map (at the expense of increased memory usage). However, there is a fundamental trade-off between performance and determinism. Even if tf.data would allow for extra buffering in its map transformation, the performance might not be as good as for the non-deterministic transformation (unless the buffering is unlimited but even then non-deterministc execution might achieve better cache locality). 

For what it is worth, most users are happy with the "fast and non-deterministic" and "slower and deterministic" options. Having said that, I think that we should be able to investigate opportunities for the decoupling the parallelism and buffer size configurations in the context of our ongoing autotuning efforts.

**Comment by zhuzilin** [comment]
@jsimsa 
Apart from the trade-off between buffer size and performance, I think one barrier of implementing the background buffer of parallel mapping is that it is not sure whether it need to process the next one. For example, we may have code like:
```python
def parse_and_preprocessing(x):
  # very slow

ds = tf.data.Dataset.TFRecordDataset("example.tfrecord")
ds = ds.repeat()
ds = ds.map(really_slow_function, num_parallel_calls=10)
ds = ds.take(3)
```
In this snippet, the map does not need to continue to process more than 3 elements and make it continue to run in background may be wasteful. For this specific problem, we may add a grappler pass to hoist the `take`. But the main idea is the dataset op has no idea what will happen after it. I believe some systematic graph optimization for the data pipeline would help. I wonder if this is among the team's interests and I'd love to contribute :).

**Comment by jsimsa** [comment]
@zhuzilin the inefficiency related to asychronously pre-computing more data than is needed is not specific to this issue and applies to any of the existing asynchronous tf.data transformations (parallel map, parallel interleave, fused map + batch, and prefetch).

Pushing transformations that discard data (such as `skip`, `take`, and `shard`) as close to the data source as possible through graph rewrites makes sense and the tf.data team would welcome your contributions to that end. To get started, I suggest you take a look at existing tf.data [graph rewrites](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/grappler/optimizers/data), such as the [no-op elimination](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/data/noop_elimination.cc) optimization.

**Comment by zhuzilin** [comment]
@jsimsa Great! Thank you for your advice and I'll start working on that.

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210531, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/ca03b425f5ddd67bbe6e71cdfeb3b447/39992.ipynb). Thanks!

**Comment by mohantym** [comment]
Hi @eriikj ! 
We are checking to see whether you still need help in this issue. Have you checked above [suggestion](https://github.com/tensorflow/tensorflow/issues/39992#issuecomment-642797494) yet?
